src/Microsoft.DotNet.XHarness.iOS.Shared/Listeners/TunnelBore.cs

```diff
@@ -0,0 +1,69 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.DotNet.XHarness.iOS.Shared.Execution;
+using Microsoft.DotNet.XHarness.iOS.Shared.Logging;
+
+namespace Microsoft.DotNet.XHarness.iOS.Shared.Listeners {
+
+	// manages the tunnels that are used to communicate with the devices. We want to create a single tunnel per device
+	// since we can only run one app per device, this should not be a problem.
+	//
+	// definition: 
+	// A tunnel boring machine, also known as a "mole", is a machine used to excavate tunnels with a circular cross section
+	// through a variety of soil and rock strata. They may also be used for microtunneling. 
+	public interface ITunnelBore : IDisposable {
+
+		// create a new tunnel for the device with the given name.
+		ITcpTunnel Create (string device, ILog mainLog);
+
+		// close a given tunnel
+		Task Close (string device);
+	}
+
+	public class TunnelBore : ITunnelBore {
+
+		readonly object tunnelsLock = new object ();
+		readonly IProcessManager processManager;
+		readonly Dictionary<string, TcpTunnel> tunnels = new Dictionary<string, TcpTunnel> ();
+
+		public TunnelBore (IProcessManager processManager)
+		{
+			this.processManager = processManager ?? throw new ArgumentNullException (nameof (processManager));
+		}
+
+		// Creates a new tcp tunnel to the given device that will use the port from the passed listener. 
+		public ITcpTunnel Create (string device, ILog mainLog)
+		{
+			lock (tunnelsLock) {
```

Comment:
I think you could simplify things and make them more safe by using `ConcurrentDictionary` for `tunnels`